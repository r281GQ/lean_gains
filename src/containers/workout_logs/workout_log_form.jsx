import React, { Component } from 'react';
import { connect } from 'react-redux';
import { FlatButton } from 'material-ui';
import { DatePicker } from 'redux-form-material-ui';
import {
  Field,
  FieldArray,
  reduxForm,
  formValueSelector
} from 'redux-form/immutable';
import moment from 'moment';

import {
  createWorkoutLog,
  updateWorkoutLog
} from './../../store/actionCreators/workout_log_action_creators';
import ExerciseFieldArray from './../../components/exercies_field_array';

class WorkoutLogFormContainer extends Component {
  componentWillMount() {
    console.log(this.props);
    if (this.props.type !== 'createBefore')
      this.props.change('exercises', this.props.defaultValue);
    this.props.change('date', moment().toDate());
  }

  _disableThese = disabledDates => date =>
    _.find(disabledDates, disabledDate =>
      moment(disabledDate).isSame(date, 'day')
    )
      ? true
      : false;

  render() {
    console.log(this.props);
    let { createWorkoutLog, handleSubmit, updateWorkoutLog } = this.props;
    return (
      <div>
        <form
          onSubmit={handleSubmit(
            ({ date, exercises }) =>
              this.props.type === 'edit'
                ? updateWorkoutLog({
                    exercises,
                    _id: this.props.match.params.id
                  })
                : createWorkoutLog({ exercises, date })
          )}
        >
          <div>
            {this.props.type === 'createBefore'
              ? <div>
                  <Field
                    name="date"
                    component={input =>
                      <DatePicker
                        {...input}
                        maxDate={moment().toDate()}
                        minDate={moment().subtract(110, 'years').toDate()}
                        formatDate={value => moment(value).format('DD-MM-YYYY')}
                        shouldDisableDate={this._disableThese(
                          this.props.datesWithWorkoutLogs.toJS()
                        )}
                      />}
                  />{' '}
                </div>
              : null}

            <FieldArray
              name="exercises"
              component={ExerciseFieldArray}
            />
            <FlatButton
              type="submit"
              label={this.props.match.params.id ? 'modify' : 'create'}
            />
          </div>
        </form>
      </div>
    );
  }
}

const selector = formValueSelector('workoutLog');

//TODO own props needs to be moved to the bottom connect
// WorkoutLogFormContainer = connect(
//   state => ({
//     datesWithWorkoutLogs: state.getIn(['workoutLogs', 'dates']),
//     exercise1s: selector(state, 'exercises')
//   }),
//   dispatch => ({
//     createWorkoutLog: workoutLog => dispatch(createWorkoutLog(workoutLog)),
//     updateWorkoutLog: workoutLog => dispatch(updateWorkoutLog(workoutLog))
//   })
// )(WorkoutLogFormContainer);

export default connect(
  state => ({
    datesWithWorkoutLogs: state.getIn(['workoutLogs', 'dates'])
  })

,
  dispatch => ({
    createWorkoutLog: workoutLog => dispatch(createWorkoutLog(workoutLog)),
    updateWorkoutLog: workoutLog => dispatch(updateWorkoutLog(workoutLog))
  })

)(
  reduxForm({
    form: 'workoutLog'
  })(WorkoutLogFormContainer)
);
