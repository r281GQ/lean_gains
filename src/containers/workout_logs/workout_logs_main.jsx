import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

import {
  setSelectedMonthForWorkoutLogs,
  openWorkoutModal,
  closeWorkoutModal,
  setSelectedWorkoutLog
} from './../../store/actionCreators/app_action_creators';
import {
  getWorkoutLogsForMonth,
  deleteWorkoutLog,
  getWorkoutLogDates
} from './../../store/actionCreators/workout_log_action_creators';
import monthsWithWorkoutLogs from './../../store/selectors/month_workout_log';
import workoutLogsForMonth from './../../store/selectors/workout_log_selector';
import isTodaysWorkoutLogExists from './../../store/selectors/today_log';

import CreateButton from './../../components/create_button';
import CreateButtonMinified from './../../components/create_button_minified';
import ConfirmDelete from './../../components/confirm_delete';
import CardListLog from './../../components/card_list_log';
import DateSelector from './../../components/date_selector';
import LoadingScreen from './../../components/loading';

const renderMainScreen = ({
  monthsWithWorkoutLogs,
  selectedMonth,
  getWorkoutLogsForMonth,
  setSelectedMonthForWorkoutLogs,
  isModalOpen,
  closeWorkoutModal,
  deleteWorkoutLog,
  workoutLogsForMonth,
  setSelectedWorkoutLog,
  openWorkoutModal,
  isTodaysWorkoutLogExists,
  selectedWorkoutLog
}) =>
  <div>
    <DateSelector
      months={monthsWithWorkoutLogs.toJS()}
      selectedMonth={selectedMonth}
      fetchDataForSelectedMonth={getWorkoutLogsForMonth}
      setSelectedMonth={setSelectedMonthForWorkoutLogs}
    />

    <ConfirmDelete
      title="Sure you want to delete this log?"
      isOpen={isModalOpen}
      close={closeWorkoutModal}
      deleteActions={[() => deleteWorkoutLog(selectedWorkoutLog)]}
    />

    <CardListLog
      workoutLogs={workoutLogsForMonth.toJS()}
      editLink="/workoutlogs/edit/"
      setSelectedItem={setSelectedWorkoutLog}
      onModalStateChange={openWorkoutModal}
    />

    <CreateButtonMinified link="/workoutlogs/create/before" />

    <CreateButton link="/workoutlogs/create" disabled={isTodaysWorkoutLogExists} />
  </div>;

class WorkoutLogsMainContainer extends PureComponent {
  componentWillMount = () => {
    if (this.props.datesWithWorkoutLogs.isEmpty())
      this.props.getWorkoutLogDates();
    if (this.props.workoutLogsForMonth.isEmpty())
      this.props.getWorkoutLogsForMonth(moment().format('MM-YYYY'));
    this.props.setSelectedMonthForWorkoutLogs(
      this.props.monthsWithWorkoutLogs.last()
    );
  };

  componentWillReceiveProps = (nextProps, nextState) =>
    this.props.monthsWithWorkoutLogs.isEmpty()
      ? this.props.setSelectedMonthForWorkoutLogs(
          nextProps.monthsWithWorkoutLogs.last()
        )
      : null;

  render = () =>
    this.props.isLoading ? <LoadingScreen /> : renderMainScreen(this.props);
}

const mapStateToProps = state => ({
  isLoading: state.getIn(['app', 'isLoading']),
  monthsWithWorkoutLogs: monthsWithWorkoutLogs(state),
  isTodaysWorkoutLogExists: isTodaysWorkoutLogExists(state),
  workoutLogsForMonth: workoutLogsForMonth(state),
  datesWithWorkoutLogs: state.getIn(['workoutLogs', 'dates']),
  selectedMonth: state.getIn(['app', 'selectedMonthForWorkoutLogs']),
  isModalOpen: state.getIn(['app', 'isWorkoutLogModalOpen']),
  selectedWorkoutLog: state.getIn(['app', 'selectedWorkoutLog'])
});

const mapDispatchToProps = dispatch => ({
  getWorkoutLogsForMonth: month => dispatch(getWorkoutLogsForMonth(month)),
  getWorkoutLogDates: () => dispatch(getWorkoutLogDates()),
  setSelectedMonthForWorkoutLogs: month =>
    dispatch(setSelectedMonthForWorkoutLogs(month)),
  deleteWorkoutLog: _id => dispatch(deleteWorkoutLog(_id)),
  openWorkoutModal: () => dispatch(openWorkoutModal()),
  closeWorkoutModal: () => dispatch(closeWorkoutModal()),
  setSelectedWorkoutLog: _id => dispatch(setSelectedWorkoutLog(_id))
});

export default connect(mapStateToProps, mapDispatchToProps)(
  WorkoutLogsMainContainer
);
