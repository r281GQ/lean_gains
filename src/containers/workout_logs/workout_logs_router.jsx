import React from 'react';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { connect } from 'react-redux';
import { Route } from 'react-router-dom';
import { Map } from 'immutable';

import WorkoutLogsMainContainer from './workout_logs_main';
import WorkoutLogFormContainer from './workout_log_form';
import isTrainingDay from './../../store/selectors/exercises';

import withConfirmDeleteModal from './../enhancers/confirm_delete_modal';

const WorkoutLogsRouter = ({ exercises, workoutLogs }) =>
  <div>
    <Route
      exact
      path={'/app/workoutlogs'}
      component={withConfirmDeleteModal(
        WorkoutLogsMainContainer,
        `Are you sure you want to delete this workout log?`,
        'workoutLog',
      )}
    />

    <Route
      exact
      path={'/app/workoutlogs/create/before'}
      render={props => {
        return <WorkoutLogFormContainer {...props} type={`createBefore`} />;
      }}
    />

    <Route
      exact
      path="/app/workoutlogs/edit/:id"
      render={props => {
        const defaultValue = workoutLogs
          .find((value, key) => props.match.params.id === key)
          // .get('exercises');
;
        return (
          <WorkoutLogFormContainer
            {...props}
            defaultValue={defaultValue}

          />
        );
      }}
    />

    <Route
      exact
      path={'/app/workoutlogs/create'}
      render={props => {
        const defaultValue = exercises.map(value =>
          Map().withMutations(map =>
            map
              .set(
                'name',
                value !== 'You do not have any exercise for today' ? value : '',
              )
              .set('note', '')
              .set('marker', false)
              .set('sets', Map()),
          ),
        );

        return (
          <WorkoutLogFormContainer
            {...props}
            type={`create`}
            defaultValue={defaultValue}
          />
        );
      }}
    />
  </div>;

WorkoutLogsRouter.propTypes = {
  exercises: ImmutablePropTypes.list,
  workoutLogs: ImmutablePropTypes.map,
};

export default connect(state => ({
  workoutLogs: state.getIn(['workoutLogs', 'data']),
  exercises: isTrainingDay('main')(state),
}))(WorkoutLogsRouter);
