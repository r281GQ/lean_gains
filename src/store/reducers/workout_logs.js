import { fromJS, Map, Set } from 'immutable';
import * as _ from 'lodash';

import * as workoutLogs from './../actions/workout_logs_actions';

const INITIAL_STATE = Map().withMutations(map =>
  map.set('data', Map()).set('dates', Set())
);

const reducer = (state = INITIAL_STATE, { type, payload }) => {
  switch (type) {
    case workoutLogs.WRITE_WORKOUT_LOG_DATES:
      return state.update('dates', set => set.concat(fromJS(payload)));
    case workoutLogs.WRITE_WORKOUT_LOG_DATE:
      return state.update('dates', set => set.add(payload));
    case workoutLogs.DELETE_WORKOUT_LOG_DATE:
      return state.update('dates', set => set.remove(payload));
    case workoutLogs.DELETE_WORKOUT_LOG:
      return state.deleteIn(['data', payload]);
    case workoutLogs.WRITE_WORKOUT_LOGS:
      return state.update('data', map =>
        map.concat(fromJS(_.keyBy(payload, '_id')))
      );
    case workoutLogs.WRITE_WORKOUT_LOG:
      return state.setIn(['data', payload._id], fromJS(payload));
    default:
      return state;
  }
};

export default reducer;
