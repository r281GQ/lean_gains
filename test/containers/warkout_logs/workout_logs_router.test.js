import React from 'react';
import { shallow } from 'enzyme';
import { List, Map } from 'immutable';
import { PureWorkoutLogsRouter } from './../../../src/containers/workout_logs/workout_logs_router';

describe('WorkoutLogsRouter test', () => {
  it('should render', () => {
    const component = shallow(<PureWorkoutLogsRouter />);

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  it('should create the props form of exercise from raw input for create', () => {
    const component = shallow(<PureWorkoutLogsRouter />);

    const instance = component.instance();

    const mapped = instance._mapExercisesToCreateForm(List().push('exercise'));

    expect(mapped).toEqual(
      List().push(
        Map().withMutations(map =>
          map
            .set('name', 'exercise')
            .set('note', '')
            .set('marker', false)
            .set('sets', Map())
        )
      )
    );

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  it('should create the props form of exercise from raw input for create', () => {
    const component = shallow(
      <PureWorkoutLogsRouter match={{ params: { id: 1 } }} />
    );

    const instance = component.instance();

    const mapped = instance._mapExercisesToCreateForm(
      List().push('You do not have any exercise for today')
    );

    expect(mapped).toEqual(
      List().push(
        Map().withMutations(map =>
          map
            .set('name', '')
            .set('note', '')
            .set('marker', false)
            .set('sets', Map())
        )
      )
    );
    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  it('should create the props form of exercise from raw input for edit', () => {
    const component = shallow(<PureWorkoutLogsRouter />);

    const instance = component.instance();

    const mapped = instance._mapExercisesToEditForm(
      Map().set('1', Map().set('value', 'v')),
      { match: { params: { id: '1' } } }
    );

    expect(mapped).toEqual(Map().set('value', 'v'));
    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });
});
