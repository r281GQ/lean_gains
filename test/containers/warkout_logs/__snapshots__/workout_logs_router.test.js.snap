// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkoutLogsRouter test should create the props form of exercise from raw input for create 1`] = `
<div>
  <Route
    component={[Function]}
    exact={true}
    path="/app/workoutlogs"
  />
  <Route
    exact={true}
    path="/app/workoutlogs/create/before"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/app/workoutlogs/edit/:id"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/app/workoutlogs/create"
    render={[Function]}
  />
</div>
`;

exports[`WorkoutLogsRouter test should create the props form of exercise from raw input for create 2`] = `
<div>
  <Route
    component={[Function]}
    exact={true}
    path="/app/workoutlogs"
  />
  <Route
    exact={true}
    path="/app/workoutlogs/create/before"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/app/workoutlogs/edit/:id"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/app/workoutlogs/create"
    render={[Function]}
  />
</div>
`;

exports[`WorkoutLogsRouter test should create the props form of exercise from raw input for edit 1`] = `
<div>
  <Route
    component={[Function]}
    exact={true}
    path="/app/workoutlogs"
  />
  <Route
    exact={true}
    path="/app/workoutlogs/create/before"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/app/workoutlogs/edit/:id"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/app/workoutlogs/create"
    render={[Function]}
  />
</div>
`;

exports[`WorkoutLogsRouter test should render 1`] = `
<div>
  <Route
    component={[Function]}
    exact={true}
    path="/app/workoutlogs"
  />
  <Route
    exact={true}
    path="/app/workoutlogs/create/before"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/app/workoutlogs/edit/:id"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/app/workoutlogs/create"
    render={[Function]}
  />
</div>
`;
